### ISSUE TITLE
Make AnyProvider / ErasedDataProvider be zero-alloc and cloneable

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#1479]
With the upcoming CrabBake, ErasedDataProvider, which I intend to rename to AnyProvider, will play a more central role in the data provider machinery.  For more information, see #1471.

Currently, the `upcast` method creates a `Box<ErasedDataStruct>`, which is basically a glorified `Box<dyn Any>`.  However, we do not want to require that a `Box` is created in order to make use of `AnyProvider`.

Previously, we used `DataMarker::Cart` to wrap the struct into a `dyn ErasedDataStruct` and then point to it from the yoke:

https://github.com/unicode-org/icu4x/blob/8e66c5ea835b43a6a46e9b994b31b6d92f224aed/provider/core/src/erased.rs#L68

We can still do this, *if* we promote AnyProvider to be a first-class citizen like BufferProvider.  Here's what I mean.  Currently, DataProvider has only two types of carts: a buffer, or None.  If we re-added the DataPayloadInner enum, we could allow DataProvider to have a third cart type out of the box, which would be `dyn Any` (or `dyn ErasedDataStruct`).

Needs approval from:

- [ ] @Manishearth

**Comment by sffc** [comment]
Um, right after I posted this I realized that `DataMarker::Cart` still required an allocation, just in a different place (the cart rather than the yokeable).  `Rc<dyn Any>`.  I would note that this solves the cloning problem at least, since the current solution results in payloads that are not cloneable.

Any other ideas?

**Comment by sffc** [comment]
Here's something that might work.  In CrabBake, we can get `&'static T` references to all data structs.  We can cast the `&'static T` to `&'static dyn Any`, and then yoke that to an empty cart.

**Comment by Manishearth** [comment]
Yeah for erased cases I feel like we can just project the yoke payload to/from dyn any?

**Comment by Manishearth** [comment]
I feel like a dyn Any cart is ultimately fine to add, but I'd like to avoid it if possible

**Comment by sffc** [comment]
Now that we're getting more serious about CrabBake, the question is, how do we structure AnyMarker/AnyProvider (aka ErasedData[Struct/Provider]) to best serve that use case?

Prior to #1278, erased data was basically stored like this:

```rust
// Type A: Reference to a trait object in the cart
Yoke<&'static dyn Any, Rc<dyn Any>>
```

After #1278, it was essentially changed to this:

```rust
// Type B: Boxed trait object with empty cart
Yoke<Box<dyn Any>, ()>
```

I think Type A is the better option.  If the object is statically allocated, like in CrabBake, we can still have an empty cart, and if the object is dynamically allocated, we stick it in a `Rc`, which allows us to cheaply clone it.

We would support three types of carts in DataPayload:

- Empty cart
- `Rc<[u8]>`
- `Rc<dyn Any>`

**Comment by sffc** [comment]
I should clarify what I mean about `dyn Any`.  We are actually talking about two different things that we anyify:

1. `M::Yokeable` (for CrabBake)
2. `Yoke<M::Yokeable, _>` (for the dynamic case), or `DataPayload<M>` itself

Previously, the ErasedDataStruct was always the yoke itself cast to `dyn Any`.  Downcasting the payload involved downcasting the yoke.  But now with CrabBake, we can actually cast the object itself to `dyn Any`.

It should also be theoretically possible to get a `&dyn Any` to the yokeable part of a cart that is stuck into a `dyn Any`, but we haven't successfully done that before.  It would basically mean calling `.get()` on the yoke, erasing its lifetimes with the Yokeable crate, upcasting the reference to `dyn Any`, and then sticking the yoke into a `Rc<dyn Anyn>`.

**Comment by sffc** [comment]
It should be possible to write a safe helper function in the Yoke crate to get from `Rc<Yoke<Y,  C>>` to `Yoke<&'static dyn Any, Rc<dyn Any>>`.

1. Cast the `Rc<Yoke<Y, C>>` to `Rc<dyn Any>`.
2. Pass it as the cart to `Yoke::attach_to_cart`.
3. In the closure, downcast `Rc<dyn Any>` back to `Rc<Yoke<Y, C>>`.
4. Call `.get()` to get a `&'a Y<'a>`.
5. Use `Y::make()` to get a `&'a Y<'static>`.
6. Cast that to a `&'a dyn Any` and return it from the closure.

**Comment by sffc** [comment]
I think I'm also okay decoupling these two questions.  We can make `AnyMarker` refer to `&'static dyn Any` which is a reference to an upcast `M::Yokeable`, which solves the CrabBake use case, and then in the future we can add the `Rc<dyn Any>` cart in order to support dynamically allocated objects with borrows.

**Comment by sffc** [comment]
Note: downcast would become project with ZeroCopyFrom.

**Comment by sffc** [comment]
@Manishearth Are you concerned at all about the fact that this strategy for AnyMarker means that every time we downcast-upcast we essentially add an extra layer of nesting to the Yoke?

For example:

```rust
// Start
Yoke<Y, Rc<[u8]>>

// Upcast 1
Yoke<&'static dyn Any, ErasedRcCart> // perceived
Yoke<&'static Y, Yoke<Y, Rc<[u8]>>> // actual

// Downcast 1
Yoke<Y, ErasedRcCart> // perceived
Yoke<Y, Yoke<&'static Y, Yoke<Y, Rc<[u8]>>>> // actual

// Upcast 2
Yoke<&'static dyn Any, ErasedRcCart> // perceived
Yoke<&'static Y, Yoke<Y, Yoke<&'static Y, Yoke<Y, Rc<[u8]>>>>> // actual
```

**Comment by sffc** [comment]
If this is a problem, perhaps we could solve it by making making the cart `Rc<dyn Any>` so that we can `downcast` it back to `Rc<Yoke<...>>`, and then we can get the yoke back by calling `Rc::try_unwrap`, cloning the inner yoke if it fails.

**Comment by Manishearth** [comment]
@sffc Yes, this is why I was saying that ErasedCart won't work and we will need to use `Rc<dyn Any>`.

**Comment by sffc** [comment]
We don't even really need Yoke for AnyProvider.  We can return AnyPayload:

```rust
struct AnyPayload(Rc<dyn Any>)
```

And the downcast of AnyPayload to DataPayload is something like

```rust
impl AnyPayload {
    pub fn downcast<M>(self) -> Result<DataPayload<M>> {
        let rc_payload: Rc<DataPayload<M>> = self.downcast()?;
        match rc_payload.try_unwrap() {
            Ok(payload) => payload,
            Err(rc_payload) => rc_payload.deref().clone(),
        }
    }
}
```

Seems pretty clean?

**Comment by sffc** [comment]
I guess then we're back to the issue where we don't want to be required to wrap stuff in `Rc` if we don't need to do that (i.e. CrabBake).

I guess the solution to that problem is to do

```rust
enum AnyPayloadInner {
    AnyYoke(Rc<dyn Any>),
    StaticRef(&'static dyn Any),
}
pub struct AnyPayload(AnyPayloadInner);
```

And the `downcast` impl produces a Yoke with an empty cart for StaticRef.

**Comment by Manishearth** [comment]
I feel like AnyProvider should just return AnyMarker data payloads? CrabBake should be capable of producing "owned" Yokes. We can use projection to construct AnyMarker data payloads.

I'd rather not have enums like that if possible.

**Comment by sffc** [comment]
CrabBake is fine. But it's the Override and Cache use cases that don't really work very nicely with AnyMarker+DataPayload. If we want to avoid the infinite hidden nesting problem from my comment above, then we need to add Rc-dyn-Any as a Cart type and read from the cart when downcasting.

**Comment by Manishearth** [comment]
Oh. I see the issue. The details here changes a bit if we have a combined cart type (using `EitherCart`, etc). For clarity I'm just going to call it `CombinedCart` with `RcBuf`, `None`, and `RcAny` variants.

What if, during downcast, we check the typeid of the  `dyn Any` cart, and if it's the right type, just return the cart? Typically unwrapping carts is a bad idea since the yokeable might be modified, but in this case the yokeable is `&'static dyn Any` and thus _cannot_ be modified.

```
// NOTE: RcBuf and RcAny are not types, they are values, I use them here to hint at what the underlying value is

// Start
Yoke<Y, CombinedCart::RcBuf>

// Upcast 1: stuff yoke into an `Rc`, cast to dyn Any, cache if necessary, create new yoke
Yoke<&'static dyn Any, CombinedCart::RcAny> // perceived
Yoke<&'static Y, Rc<Yoke<Y, CombinedCart::RcBuf>>> // actual
// For clarity, we may wish to have the actual cart actually be
// `Rc<UpcastedCart<Yoke<...>>` so that there's no chance of confusion when downcasting. `UpcastedCart` can then be a private type.

// Downcast 1
// check typeid of CombinedCart. Check succeeds!
// call .into_backing_cart(), clone to extract the yoke
Yoke<Y, CombinedCart> // perceived
Yoke<Y, CombinedCart::RcBuf> // actual (back where we started!)
```

Note that this does require a clone of the yoke when extracting it from the `Rc` cart, but as far as I can tell `AnyPayload` also has that requirement.

Now let's look at what happens with AnyMarkers that are not constructed from upcasts (e.g. the impl of AnyProvider on CrabBake):

```rust
Yoke<&'static dyn Any, CombinedCart::None>

// Downcast:
// check type id of cart. check failed
// project the yoke with ZCF instead:

Yoke<Y, CombinedCart::None>
```

The nice thing about this is that AnyMarker stays interoperable, and you can construct AnyMarker data with any kind of cart.

That said, I wonder how much utility there is for AnyMarker to be truly interoperable. It might be fine if we just have AnyPayload. It seems like what I'm proposing here is essentially AnyPayload, but retrofit into Yoke using type ids. This might be a ... little bit more convenient for users? But I don't know.

**Comment by sffc** [comment]
Yes.  FWIW, what you are describing is essentially what we used to do prior to #1278.

> It seems like what I'm proposing here is essentially AnyPayload, but retrofit into Yoke using type ids. In your case you have AnyPayload's two variants, in my case I have "AnyMarker payload with correct type id in cart" and "AnyMarker payload with other typeid in cart". This might be a ... little bit more convenient for users? But I don't know.

Right. I think both are workable solutions.  If I had to state the pros and cons:

Custom AnyPayload type:

- Pro: New type with its own data model, methods, and documentation focused on this use case
- Pro: Does not add bloat to DataPayload
- Con: The new type may lead to complexity for users

Re-use DataPayload with AnyMarker:

- Pro: Re-uses the DataPayload signature and everything that goes along with it
- Con: Adds a third cart type to DataPayload that is not used by anyone else
- Con: Running different code paths based on TypeId seems fairly fragile and difficult to properly debug, test, document, and explain
- Con: Many DataPayload methods do not make sense on `DataPayload<AnyMarker>`.

> That said, I wonder how much utility there is for AnyMarker to be truly interoperable. It might be fine if we just have AnyPayload.

I think the need for interoperability is extremely limited.  Neither `AnyPayload` nor `DataPayload<AnyMarker>` are useful on their own; they should both be downcast to `DataPayload<M>`.  I tend to think that the needs of AnyPayload are sufficiently different from the needs of DataPayload that it's cleaner if we separate the two.

I don't intend to touch `BufferMarker` or `SerdeSeDataStructMarker` at this time, though that may change in the future depending on how `AnyPayload` works out.

OK?
